pyro/distributions/transforms/block_naf.py

```diff
@@ -0,0 +1,232 @@
+# This implementation is adapted in part from https://github.com/nicola-decao/BNAF under the MIT license.
+from __future__ import absolute_import, division, print_function
+
+import math
+
+import torch
+import torch.nn as nn
+from pyro.distributions.torch_transform import TransformModule
+from torch.distributions import constraints
+import torch.nn.functional as F
+
+from pyro.distributions.util import copy_docs_from
+from pyro.distributions.transforms.naf import ELUMixin, LeakyReLUMixin, TanhMixin, SigmoidalMixin
+
+eps = 1e-8
+
+
+def log_matrix_product(A, B):
+    """
+    Computes the matrix products of two matrices in log-space, returning the result in log-space.
+    This is useful for calculating the vector chain rule for Jacobian terms.
+    """
+    return torch.logsumexp(A.unsqueeze(-1) + B.unsqueeze(-3), dim=-2)
+
+
+@copy_docs_from(TransformModule)
+class BlockNAFFlow(TransformModule):
+    """
+    An implementation of Block Neural Autoregressive Flow (block-NAF) (De Cao et al., 2019). Block-NAF uses a
+    similar transformation to deep dense NAF, building the autoregressive NN into the structure of the flow,
+    in a sense.
+
+    Together with `TransformedDistribution` this provides a way to create richer variational approximations.
+
+    Example usage:
+
+    >>> base_dist = dist.Normal(torch.zeros(10), torch.ones(10))
+    >>> naf = BlockNAFFlow(input_dim=10)
+    >>> pyro.module("my_naf", naf)  # doctest: +SKIP
+    >>> naf_dist = dist.TransformedDistribution(base_dist, [naf])
+    >>> naf_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    The inverse operation is not implemented. This would require numerical inversion, e.g., using a
+    root finding method - a possibility for a future implementation.
+
+    :param input_dim: The dimensionality of the input and output variables.
+    :type input_dim: int
+    :param hidden_factors: Hidden layer i has hidden_factors[i] hidden units per input dimension. This
+        corresponds to both :math:`a` and :math:`b` in De Cao et al. (2019). The elements of hidden_factors
+        must be integers.
+    :type hidden_factors: list
+    :param activation: Activation function to use. One of 'ELU', 'LeakyReLU', 'sigmoid', or 'tanh'.
+    :type activation: string
+    :param residual: Type of residual connections to use. Choices are "None", "normal" for
+        :math:`\\mathbf{y}+f(\\mathbf{y})`, and "gated" for :math:`\\alpha\\mathbf{y} + (1 - \\alpha\\mathbf{y})`
+        for learnable parameter :math:`\\alpha`.
+    :type residual: string
+
+    References:
+
+    Block Neural Autoregressive Flow [arXiv:1904.04676]
+    Nicola De Cao, Ivan Titov, Wilker Aziz
+    """
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
```

Comment:
Should this be false given that `inverse` isn't implemented?

Reply:
The transform is bijective even though an analytic inverse doesn't exist. I think this should be true based on the comment: https://github.com/pytorch/pytorch/blob/master/torch/distributions/transforms.py#L63